Mathlib/RingTheory/AlgebraicIndependent/AlgebraicClosure.lean

```diff
@@ -0,0 +1,75 @@
+/-
+Copyright (c) 2024 Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Junyan Xu
+-/
+import Mathlib.FieldTheory.AlgebraicClosure
+import Mathlib.RingTheory.Algebraic.Integral
+import Mathlib.RingTheory.AlgebraicIndependent.Transcendental
+
+/-!
+# Algebraic independence persists to the algebraic closure
+
+## Main results
+
+* `AlgebraicIndependent.extendScalars`: if A/S/R is a tower of algebras with S/R algebraic,
+then a family of elements in A that are algebraically independent over R remains algebraically
+independent over S, provided that S has no zero divisors.
+
+* `AlgebraicIndependent.algebraicClosure`: an algebraically independent family remains
+algebraically independent over the algebraic closure.
+-/
+
+open Function Algebra
+
+namespace AlgebraicIndependent
+
+variable {ι R S A : Type*} {x : ι → A}
+variable [CommRing R] [CommRing S] [CommRing A]
+variable [Algebra R S] [Algebra R A] [Algebra S A] [IsScalarTower R S A]
+variable [NoZeroDivisors S] (hx : AlgebraicIndependent R x)
+include hx
+
+theorem extendScalars [alg : Algebra.IsAlgebraic R S]
+    (inj : Injective (algebraMap S A)) : AlgebraicIndependent S x := by
+  nontriviality S
+  have := inj.nontrivial
+  refine algebraicIndependent_of_finite_type' inj fun t fin ind i hi ↦ ?_
+  let Rt := adjoin R (x '' t)
+  let St := adjoin S (x '' t)
+  let _ : Algebra Rt St :=
+    (Rt.inclusion (T := St.restrictScalars R) <| adjoin_le <| by exact subset_adjoin).toAlgebra
+  have : IsScalarTower Rt St A := .of_algebraMap_eq fun ⟨y, _⟩ ↦ show y = y from rfl
+  have : NoZeroDivisors St := (Set.image_eq_range _ _ ▸ ind.aevalEquiv)
+    |>.symm.injective.noZeroDivisors _ (map_zero _) (map_mul _)
+  have : NoZeroDivisors Rt := (Subalgebra.inclusion_injective _).noZeroDivisors
+    (algebraMap Rt St) (map_zero _) (map_mul _)
+  have : Algebra.IsAlgebraic Rt St := ⟨fun ⟨y, hy⟩ ↦ by
+    rw [← isAlgebraic_algHom_iff (IsScalarTower.toAlgHom Rt St A) Subtype.val_injective]
+    show IsAlgebraic Rt y
+    exact adjoin_induction (fun _ h ↦ isAlgebraic_algebraMap (⟨_, subset_adjoin h⟩ : Rt))
+      (fun z ↦ ((alg.1 z).algHom (IsScalarTower.toAlgHom R S A)).extendScalars fun _ _ eq ↦ by
+        exact hx.algebraMap_injective congr($eq.1)) (fun _ _ _ _ ↦ .add) (fun _ _ _ _ ↦ .mul) hy⟩
+  show Transcendental St (x i)
+  exact (hx.transcendental_adjoin hi).extendScalars Subtype.val_injective
+
+theorem extendScalars_of_isIntegral [Algebra.IsIntegral R S]
+    (inj : Injective (algebraMap S A)) : AlgebraicIndependent S x := by
+  nontriviality S
+  have := Module.nontrivial R S
+  exact hx.extendScalars inj
+
+theorem subalgebraAlgebraicClosure [IsDomain R] [NoZeroDivisors A] :
+    AlgebraicIndependent (Subalgebra.algebraicClosure R A) x :=
+  hx.extendScalars Subtype.val_injective
+
+protected theorem integralClosure [NoZeroDivisors A] :
+    AlgebraicIndependent (integralClosure R A) x :=
```

Comment:
Here, I think `integralClosure` can be replaced by any `Subalgebra R A` that `IsAlgebraic`.

Reply:
Added two `subalgebra` versions in [ca2c60c](https://github.com/leanprover-community/mathlib4/pull/19708/commits/ca2c60c135e4aa027e24709f8679714bcd037b29); the IsIntegral version is needed for the degenerate case R=0 where nothing is algebraic over R.